### ISSUE TITLE
--insecure-registry ignored

### ISSUE THREAD

**Issue by rafalkrupinski** [docker/docker#27113]
**Description**

docker push command hangs when pushing to an insecure registry

**Steps to reproduce the issue:**
1. setup docker with insecure registry
2. push image

**Describe the results you received:**
docker push command hangs after printing
```
The push refers to a repository [tag]
585e764decf4: Preparing
```

journalctl shows
```
msg="Not continuing with push after error: context canceled"
msg="Attempting next endpoint for push after error: Get https://server:8050/v2/: http: server gave HTTP response to HTTPS client"
```

**Describe the results you expected:**
Image being pushed

**Output of `docker version`:**

```
$ docker version
Client:
 Version:      1.12.1
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   23cf638
 Built:        Thu Aug 18 05:33:38 2016
 OS/Arch:      linux/amd64

Server:
 Version:      1.12.1
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   23cf638
 Built:        Thu Aug 18 05:33:38 2016
 OS/Arch:      linux/amd64
```

**Output of `docker info`:**

```
$ docker info
Containers: 1
 Running: 1
 Paused: 0
 Stopped: 0
Images: 28
Server Version: 1.12.1
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 54
 Dirperm1 Supported: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: host null bridge overlay
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Security Options: apparmor seccomp
Kernel Version: 4.4.0-38-generic
Operating System: Ubuntu 16.04.1 LTS
OSType: linux
Architecture: x86_64
CPUs: 4
Total Memory: 7.796 GiB
Name: ubu
ID: W3Z2:MX5V:OISD:2KLH:UIK7:6Q7R:EWHY:5DP5:6VS6:2QID:CT67:LCZV
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
WARNING: No swap limit support
Insecure Registries:
 10.73.96.0/24
 127.0.0.0/8
```
also tried with specific IP and a hostname

**Comment by michelvocks** [comment]
It seems like the repository is not configured correctly. 
`http: server gave HTTP response to HTTPS client` indicates that SSL is the problem here.

**Comment by rafalkrupinski** [comment]
I didn't setup any SSL, I thought the connection would be made over HTTP since it's "insecure registry"

**Comment by michelvocks** [comment]
`https://server:8050/v2/:` it tries to connect via SSL. Insecure registry is used to skip the certificate validation check if you use for example a self-signed certificate which is not trusted by default. 

1. What is the name/tag from the image?
2. How did you setup the registry?
3. If possible, try to remove the insecure registry part (might be a conflict here).

**Comment by rafalkrupinski** [comment]
- 1 It's just reg:8050/service:03102016
- 2
```
  registry:
    image: 'registry:2.5.1'
    container_name: registry
    restart: unless-stopped
    ports:
      - '8050:5000'
    volumes:
      - 'docker-registry:/var/lib/registry'
    networks:
      - infra
```
- 3 Tried without insecure registry; tail of docker info:
Insecure Registries:
 127.0.0.0/8
$ docker push reg:8050/service:03102016
The push refers to a repository [reg:8050/service]
Get https://reg:8050/v1/_ping: http: server gave HTTP response to HTTPS client

I also tried tagging with the registry IP address, with the same result.
My colleague tried pushing to the same server (using insecure registry, same version, Windows 10 client) with success.

**Comment by michelvocks** [comment]
@rafalkrupinski Thanks for the information. I was able to reproduce your issue. 
It you change the image tag from `reg:8050/service:03102016` to `localhost:8050/service:03102016` it should work. The problem is that the `--insecure-registry` parameter was set with a wrong value. You might forgot to also add the port to the address? Here is an example how to do it: https://docs.docker.com/registry/insecure/#/deploying-a-plain-http-registry

**Comment by rafalkrupinski** [comment]
@michelvocks The registry resides on a remote machine, so obviously I couldn't use localhost to access it. Pushing to localhost should always work, since it's the default insecure registry.

I tried reproducing the problem on my machine and I couldn't...
 - I've created the registry using the same compose file.
 - Added the containers IP to /etc/hosts
 - tagged and pushed an image
 - ...success

Still can't connect to the remote registry ...even after reinstalling the docker-engine.
I'd appreciate any other hints.
